File: c-sharp/sem01/j05/sql_backend/DAO/CapteurDAO.cs
using Microsoft.Data.SqlClient;
using sql_backend.Model;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
// DAO = Data Access Object
// Gère la communication avec SQL Server (CRUD).
// CRUD = Create (créer), Read (lire), Update (mettre à jour) et Delete (supprimer)

namespace sql_backend.DAO
{
    public class CapteurDAO
    {
        private string _connectionString;

        public CapteurDAO(string connectionString)
        {
            _connectionString = connectionString;
        }

        public void Create()
        {
            Capteur cap = new Capteur();
            cap.Id = 0;
            Console.WriteLine("Enter name:");
            cap.Nom = Console.ReadLine();
            Console.WriteLine("Enter type:");
            cap.Type = Console.ReadLine();
            Console.WriteLine("Enter unit:");
            cap.Unite = Console.ReadLine();

            using (SqlConnection conn = new SqlConnection(_connectionString))
            {
                conn.Open();
                string sql = "INSERT INTO Capteurs (nom, type, unite) VALUES ('"+ cap.Nom.Replace("\'", "\'\'") + "', '"+ cap.Type.Replace("\'", "\'\'") + "', '"+ cap.Unite.Replace("\'", "\'\'") + "')";
                SqlCommand cmd = new SqlCommand(sql, conn);
                cmd.ExecuteNonQuery();
                Console.WriteLine("Connexion OK - SQL entry created");
            }
        }
        public List<Capteur> FindAll()
        {
            List<Capteur> mes_cap = new List<Capteur>(); // Liste vide de capteurs
            using (SqlConnection conn = new SqlConnection(_connectionString))
            {
                conn.Open();

                string sql = "SELECT id, nom, type, unite FROM Capteurs";
                SqlCommand cmd = new SqlCommand(sql, conn);
                SqlDataReader reader = cmd.ExecuteReader();

                while (reader.Read())
                {
                    Capteur cap = new Capteur();
                    cap.Id = Convert.ToInt32(reader["id"]);
                    cap.Nom = Convert.ToString(reader["nom"]);
                    cap.Type = Convert.ToString(reader["type"]);
                    cap.Unite = Convert.ToString(reader["unite"]);
                    mes_cap.Add(cap);
                }
            }
            return mes_cap;
        }
        public Capteur findById(int id)
        {
            Capteur cap = new Capteur();

            using (SqlConnection conn = new SqlConnection(_connectionString))
            {
                conn.Open();
                string sql = "SELECT * FROM Capteurs WHERE Id="+id;
                SqlCommand cmd = new SqlCommand(sql, conn);
                SqlDataReader reader = cmd.ExecuteReader();

                while (reader.Read())
                {
                    cap.Id = Convert.ToInt32(reader["id"]);
                    cap.Nom = Convert.ToString(reader["nom"]);
                    cap.Type = Convert.ToString(reader["type"]);
                    cap.Unite = Convert.ToString(reader["unite"]);
                }
            }
            return cap;
        }
        public void Update()
        {
            Console.WriteLine("Which Id should be modified?");
            int MonId = Convert.ToInt32(Console.ReadLine());
            Console.WriteLine("Searched index:"+MonId);
            Capteur cap = findById(MonId);
            cap.display();

            Console.WriteLine("Enter name (hit enter to keep existing):");
            string s = Console.ReadLine();
            if (s != "") cap.Nom = s;
            Console.WriteLine("Enter type (hit enter to keep existing):");
            s = Console.ReadLine();
            if (s != "") cap.Type = s;
            Console.WriteLine("Enter unit (hit enter to keep existing):");
            s = Console.ReadLine();
            if (s != "") cap.Unite = s;

            cap.display();
            using (SqlConnection conn = new SqlConnection(_connectionString))
            {
                conn.Open();
                string sql = "UPDATE Capteurs SET nom = '" + cap.Nom.Replace("\'", "\'\'") 
                    + "', type = '" + cap.Type.Replace("\'", "\'\'") 
                    + "', unite = '" + cap.Unite.Replace("\'", "\'\'") 
                    + "'  WHERE id = " + cap.Id ;
                SqlCommand cmd = new SqlCommand(sql, conn);
                cmd.ExecuteNonQuery();
                Console.WriteLine("Connexion OK - SQL entry modified");
            }
        }
        public void Delete()
        {
            Console.WriteLine("Which Id should be deleted?");
            int MonId = Convert.ToInt32(Console.ReadLine());
            Console.WriteLine("Searched index:" + MonId);
            Capteur cap = findById(MonId);
            cap.display();

            Console.WriteLine("Confirm deletion (y/n):");
            string s = Console.ReadLine();
            if (s == "y") {
                using (SqlConnection conn = new SqlConnection(_connectionString))
                {
                    conn.Open();
                    string sql = "DELETE FROM Capteurs WHERE id = " + cap.Id;
                    SqlCommand cmd = new SqlCommand(sql, conn);
                    cmd.ExecuteNonQuery();
                    Console.WriteLine("Connexion OK - SQL entry deleted");
                }
            }
        }
    }

}

File: java/sem01/j05/sql_backend/src/dao/CapteurDAO.java
// Un DAO (Data Access Object) est une classe spécialisée dans l'accès aux données
// Gère la communication avec SQL Server (CRUD).
// CRUD = Create (créer), Read (lire), Update (mettre à jour) et Delete (supprimer)

package dao;
import model.Capteur;
import java.sql.*;
import java.util.LinkedList;

public class CapteurDAO {
    private final String _connectionString;

    public CapteurDAO(String connectionString) {
        _connectionString = connectionString;
    }

    public void Insert(Capteur c) {
        try (Connection conn = DriverManager.getConnection(_connectionString)) {
            System.out.println("Connexion OK");
            String sql = "INSERT INTO Capteurs (nom, type, unite) VALUES (?, ?, ?)";
            PreparedStatement ps = conn.prepareStatement(sql);
            ps.setString(1, c.getNom());
            ps.setString(2, c.getType());
            ps.setString(3, c.getUnite());
            ps.executeUpdate();
            System.out.println("Connexion OK - SQL entry created");
        } catch (SQLException e) {
            System.out.println("SQL Error");
        }
    }

    public LinkedList<Capteur> FindAll() {
        LinkedList<Capteur> mesCap = new LinkedList<>(); // Liste vide de capteurs
        Connection conn = null;
        Statement stmt = null;
        ResultSet rs = null;
        try {
            // Établir la connexion
            conn = DriverManager.getConnection(_connectionString);
            System.out.println("Connexion OK");

            // Créer une requête
            stmt = conn.createStatement();
            String sql = "SELECT id, nom, type, unite FROM Capteurs";

            // Exécuter la requête
            rs = stmt.executeQuery(sql);

            // Lire les résultats
            while (rs.next()) {
                Capteur cap = new Capteur();
                cap.setId(rs.getInt("id"));
                cap.setNom(rs.getString("nom"));
                cap.setType(rs.getString("type"));
                cap.setUnite(rs.getString("unite"));
                mesCap.addLast(cap);
            }
        } catch (SQLException e) {
            System.out.println("SQL Error");
        } finally {
            // Fermer les ressources
            try {
                if (rs != null) rs.close();
                if (stmt != null) stmt.close();
                if (conn != null) conn.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
        System.out.println("Connexion OK - SQL read complete");
        return mesCap;
    }

    public Capteur findById(int id) {
        Capteur cap = new Capteur();
        Connection conn = null;
        Statement stmt = null;
        ResultSet rs = null;
        try {
            // Établir la connexion
            conn = DriverManager.getConnection(_connectionString);
            System.out.println("Connexion OK");

            // Créer une requête
            stmt = conn.createStatement();
            String sql = "SELECT * FROM Capteurs WHERE Id=" + id;
            System.out.println(sql);

            // Exécuter la requête
            rs = stmt.executeQuery(sql);

            // Lire les résultats
            while (rs.next()) {
                cap.setId(rs.getInt("id"));
                cap.setNom(rs.getString("nom"));
                cap.setType(rs.getString("type"));
                cap.setUnite(rs.getString("unite"));
            }
        } catch (SQLException e) {
            System.out.println("SQL Error");
            e.printStackTrace();
        } finally {
            // Fermer les ressources
            try {
                if (rs != null) rs.close();
                if (stmt != null) stmt.close();
                if (conn != null) conn.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
        System.out.println("Connexion OK - SQL read complete");
        return cap;
    }

    public void Update(Capteur c) {
        Connection conn = null;
        Statement stmt = null;
        try {
            // Établir la connexion
            conn = DriverManager.getConnection(_connectionString);
            System.out.println("Connexion OK");

            // Créer une requête
            stmt = conn.createStatement();
            String sql = "UPDATE Capteurs SET nom = '" + c.getNom().replace("'", "''")
                    + "', type = '" + c.getType().replace("'", "''")
                    + "', unite = '" + c.getUnite().replace("'", "''")
                    + "'  WHERE id = " + c.getId();
            System.out.println(sql);

            // Exécuter la requête
            stmt.executeUpdate(sql);
        } catch (SQLException e) {
            System.out.println("SQL Error");
            e.printStackTrace();
        } finally {
            // Fermer les ressources
            try {
                if (stmt != null) stmt.close();
                if (conn != null) conn.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
        System.out.println("Connexion OK - SQL entry modified");
    }

    public void Delete(int id) {
        Connection conn = null;
        Statement stmt = null;
        try {
            // Établir la connexion
             conn = DriverManager.getConnection(_connectionString);
             System.out.println("Connexion OK");

            // Créer une requête
            stmt = conn.createStatement();
            String sql = "DELETE FROM Capteurs WHERE id = " + id;
            System.out.println(sql);

            // Exécuter la requête
            stmt.executeUpdate(sql);
            } catch (SQLException e) {
            System.out.println("SQL Error");
            e.printStackTrace();
        } finally {
            // Fermer les ressources
            try {
                if (stmt != null) stmt.close();
                if (conn != null) conn.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
        System.out.println("Connexion OK - SQL entry deleted");
    }

}
